add-item-request

Maxim-Shch opened the pull request:
Реализовали новую ветку запросов.
Что нового:
1) Добавляем запрос вещи:
Пользователь создаёт такой запрос, когда не может найти нужную вещь, воспользовавшись поиском, но при этом надеется, что у кого-то она всё же имеется. Другие пользователи могут просматривать подобные запросы и, если у них есть описанная вещь и они готовы предоставить её в аренду, добавлять нужную вещь в ответ на запрос.
Таким образом, вам нужно добавить четыре новых эндпоинта:
POST /requests — добавить новый запрос вещи. 
GET /requests — получить список своих запросов вместе с данными об ответах на них. 
GET /requests/all?from={from}&size={size} — получить список запросов, созданных другими пользователями. 
GET /requests/{requestId} — получить данные об одном конкретном запросе вместе с данными об ответах на него в том же формате, что и в эндпоинте GET /requests. 
2) Добавляем опцию ответа на запрос:
Добавим ещё одну полезную опцию в ваше приложение, чтобы пользователи могли отвечать на запросы друг друга. Для этого при создании вещи должна быть возможность указать id запроса, в ответ на который создаётся нужная вещь.
3) Добавляем пагинацию к существующим эндпоинтам:
Теперь вернёмся к улучшению, о котором мы упомянули ранее. Вы уже используете в запросе GET /requests/all пагинацию, поскольку запросов может быть очень много. 
4) Добавили тесты

Maxim-Shch committed:
Реализлвана новая ветка add-item-request

Maxim-Shch committed:
Исправил ошибки

Maxim-Shch committed:
Исправил ошибки

Maxim-Shch committed:
Удалены классы InMemoryItemStorage и InMemoryUserStorage

Maxim-Shch committed:
Удалены пакеты storage

Maxim-Shch committed:
Увеличено покрытие тестов в классе ItemServiceImplTest

Maxim-Shch committed:
Увеличено покрытие тестов в классе ItemServiceImplTest

Maxim-Shch committed:
Удалены неиспользуемые исключения

irek-valiev commented on src/main/java/ru/practicum/shareit/request/controller/ItemRequestController.java:
```diff
@@ -0,0 +1,49 @@
+    private final ItemRequestService requestService;
+
+    @PostMapping
+    public ItemRequestDto createNewRequest(@RequestHeader("X-Sharer-User-Id") final Long userId,
```
у тебя в трех контроллерах из четырех в хидере присутствует константа "X-Sharer-User-Id", предлагаю вынести ее в константу сервиса. Для этого делаешь отдельных класс типа Constants и в нем добавляешь статическую переменную и используешь во всех контроллерах. Понимаю, что получится класс из одного поля, но это же учебный проект)))

irek-valiev commented on src/main/java/ru/practicum/shareit/request/service/ItemRequestServiceImpl.java:
```diff
@@ -0,0 +1,92 @@
+
+    @Override
+    public ItemRequestDto getRequestById(Long userId, Long requestId) {
+        User requestor = userRepository.findById(userId).orElseThrow(() ->
```
в этот раз можно не выделять переменную requestor))

Maxim-Shch committed:
Исправил ошибки, а именно:
Был создан класс Constants, в котором реализовали статическую переменную для класса контролера ItemRequestController
Также в классе ItemRequestServiceImpl в методе getRequestById убрали переменную requestor

Maxim-Shch committed:
Исправил ошибки, а именно:
Был создан класс Constants, в котором реализовали статическую переменную для класса контролера ItemRequestController
Также в классе ItemRequestServiceImpl в методе getRequestById убрали переменную requestor

Maxim-Shch committed:
Исправил ошибки, а именно:
Был создан класс Constants, в котором реализовали статическую переменную для класса контролера ItemRequestController
Также в классе ItemRequestServiceImpl в методе getRequestById убрали переменную requestor

Maxim-Shch committed:
Исправил ошибки, а именно:
Был создан класс Constants, в котором реализовали статическую переменную для класса контролера ItemRequestController
Также в классе ItemRequestServiceImpl в методе getRequestById убрали переменную requestor

Maxim-Shch committed:
Пробная попытка пройти тесты

Maxim-Shch committed:
Откатил правки
